### PR TITLE
Add inplace transform example

### PR DESCRIPTION
**Pull request by yasahi-hpc**
This PR aims to adding an example for inplace 2D transforms. The documentation should be updated in another PR.

### CODE REVIEW

File: examples/08_inplace_FFT/08_inplace_FFT.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    shape_type<2> shape;
+    shape[0] = n0;
+    shape[1] = n1;
+
+    // Forward and backward complex to complex transform
+    // Define a 2D complex view to handle data
+    RightView2D<Kokkos::complex<double>> xc2c("xc2c", n0, n1);
+
+    // Create an unmanaged view on the same data
```
Comment (tpadioleau) [inline examples/08_inplace_FFT/08_inplace_FFT.cpp:34]:
```suggestion
    // Create unmanaged views on the same data
```

File: examples/08_inplace_FFT/08_inplace_FFT.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    RightView2D<Kokkos::complex<double>> xc2c_hat(xc2c.data(), n0, n1),
```
Comment (tpadioleau) [inline examples/08_inplace_FFT/08_inplace_FFT.cpp:35]: Actually for the complex to complex transform, we don't require unmanaged views:
- should we make them managed views ?
- be consistent with the c2r/r2c case ?

File: examples/08_inplace_FFT/08_inplace_FFT.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+int main(int argc, char *argv[]) {
+  Kokkos::initialize(argc, argv);
+  {
+    constexpr int n0 = 128, n1 = 128;
+    const Kokkos::complex<double> z(1.0, 1.0);
+
+    shape_type<2> shape;
+    shape[0] = n0;
+    shape[1] = n1;
```
Comment (tpadioleau) [inline examples/08_inplace_FFT/08_inplace_FFT.cpp:28]: Can we initialize `shape{n0, n1}` ?

File: examples/08_inplace_FFT/08_inplace_FFT.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                                 n1);
+
+    // Fill the input view with random data in complex space
+    Kokkos::fill_random(exec, xc2r, random_pool, z);
+
+    // Perform the complex to real transform
+    // [Important] You must use xc2r_hat to define the FFT shape correctly
+    KokkosFFT::irfft2(exec, xc2r, xc2r_hat);
+
+    Kokkos::fence();
```
Comment (tpadioleau) [inline examples/08_inplace_FFT/08_inplace_FFT.cpp:83]:
```suggestion
    exec.fence();
```

File: examples/08_inplace_FFT/08_inplace_FFT.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+template <std::size_t DIM>
+using axis_type = KokkosFFT::axis_type<DIM>;
+template <std::size_t DIM>
+using shape_type = KokkosFFT::shape_type<DIM>;
+
+int main(int argc, char *argv[]) {
+  Kokkos::initialize(argc, argv);
+  {
+    constexpr int n0 = 128, n1 = 128;
```
Comment (tpadioleau) [inline examples/08_inplace_FFT/08_inplace_FFT.cpp:23]: Not only this example but I am not convinced of making `n0` and `n1` `constexpr`. It makes me wonder if kokkos-fft can take advantage of it or not ?

File: examples/08_inplace_FFT/08_inplace_FFT.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    // Real to complex transform
+    // Define a 2D complex view to handle data
+    RightView2D<Kokkos::complex<double>> xr2c_hat("xr2c", n0, n1 / 2 + 1);
+
+    // Create an unmanaged view on the same data with the FFT shape,
+    // that is (n0, n1) -> (n0, n1/2+1) R2C transform
+    // The shape is incorrect from the view point of casting to real
+    // For casting, the shape should be (n0, (n0/2+1) * 2)
+    RightView2D<double> xr2c(reinterpret_cast<double *>(xr2c_hat.data()), n0,
+                             n1),
```
Comment (tpadioleau) [inline examples/08_inplace_FFT/08_inplace_FFT.cpp:55]: Should we provide helper functions to avoid users using `reinterpret_cast` ?

**Review by tpadioleau — DISMISSED** [review]
Looks almost good to me, just minor comments

File: examples/08_inplace_FFT/08_inplace_FFT.cpp (hunk 2 above)
Comment (yasahi-hpc) [reply examples/08_inplace_FFT/08_inplace_FFT.cpp:35]: That is a good point. 
I was just thinking of consistency with c2r and r2c.
Probably better to reuse views for simplicity.

File: examples/08_inplace_FFT/08_inplace_FFT.cpp (hunk 3 above)
Comment (yasahi-hpc) [reply examples/08_inplace_FFT/08_inplace_FFT.cpp:28]: True. I will fix that

File: examples/08_inplace_FFT/08_inplace_FFT.cpp (hunk 4 above)
Comment (yasahi-hpc) [reply examples/08_inplace_FFT/08_inplace_FFT.cpp:83]: Sure

File: examples/08_inplace_FFT/08_inplace_FFT.cpp (hunk 5 above)
Comment (yasahi-hpc) [reply examples/08_inplace_FFT/08_inplace_FFT.cpp:23]: No. Just `const` is good enough

File: examples/08_inplace_FFT/08_inplace_FFT.cpp (hunk 6 above)
Comment (yasahi-hpc) [reply examples/08_inplace_FFT/08_inplace_FFT.cpp:55]: That may be ideal. 
Can I make it another PR?

File: examples/08_inplace_FFT/08_inplace_FFT.cpp (hunk 3 above)
Comment (yasahi-hpc) [reply examples/08_inplace_FFT/08_inplace_FFT.cpp]: Actually, this variable is unnecessary. I have removed this

File: common/src/KokkosFFT_Extents.hpp [hunk 7]
```diff
@@ -33,7 +33,7 @@ namespace Impl {
 template <typename InViewType, typename OutViewType, std::size_t DIM = 1>
 auto get_extents(const InViewType& in, const OutViewType& out,
                  axis_type<DIM> axes, shape_type<DIM> shape = {},
-                 bool is_inplace = false) {
+                 [[maybe_unused]] bool is_inplace = false) {
```
Comment (tpadioleau) [inline common/src/KokkosFFT_Extents.hpp:36]: Why does it become needed only in this PR ?

File: examples/08_inplace_FFT/08_inplace_FFT.cpp [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+using execution_space = Kokkos::DefaultExecutionSpace;
+
+template <typename T>
+using RightView2D = Kokkos::View<T **, Kokkos::LayoutRight, execution_space>;
+
+template <std::size_t DIM>
+using axis_type = KokkosFFT::axis_type<DIM>;
+template <std::size_t DIM>
+using shape_type = KokkosFFT::shape_type<DIM>;
```
Comment (tpadioleau) [inline examples/08_inplace_FFT/08_inplace_FFT.cpp:18]:
```suggestion
```

**Review by tpadioleau — CHANGES_REQUESTED** [review]

File: examples/08_inplace_FFT/08_inplace_FFT.cpp (hunk 3 above)
Comment (tpadioleau) [reply examples/08_inplace_FFT/08_inplace_FFT.cpp]: This was not detected by the warnings as errors ?

**Comment by tpadioleau** [comment]
Out of curiosity, I guess there is no equivalent in scipy ? It does not support inplace transformation ?

File: common/src/KokkosFFT_Extents.hpp (hunk 7 above)
Comment (yasahi-hpc) [reply common/src/KokkosFFT_Extents.hpp:36]: Actually, we should have encountered a compiler error due to the unused variable for C2C transform. 
For some reason, we did not get compiler errors before introducing the inplace C2C transform with the same `View`.

File: examples/08_inplace_FFT/08_inplace_FFT.cpp (hunk 8 above)
Comment (yasahi-hpc) [reply examples/08_inplace_FFT/08_inplace_FFT.cpp:18]: Sorry, I forgot to remove this as well

File: common/src/KokkosFFT_Extents.hpp (hunk 7 above)
Comment (tpadioleau) [reply common/src/KokkosFFT_Extents.hpp:36]: Unexpected indeed, but I agree that it should be added anyway.

**Comment by yasahi-hpc** [comment]
@tpadioleau Thank you for your approval. I will merge this

**Pull request merged**